### DISCUSSION TITLE
Questions about the rgb-std source

### DISCUSSION THREAD

**Discussion by cymqqqq** [RGB-WG/rgb#79] [Q&A]
Hi there, I found a lot of clone() methods in the rgb-std source code, are there any better choices to optimize the code?(remove the clone), such as the code in rgb-std/persistence/inventory.rs:
let builder =
            TransitionBuilder::blank_transition(iface.clone(), schema.clone(), iimpl.clone())
                .expect("internal inconsistency");

**Comment by dr-orlovsky** [comment]
Yes, the clones better to avoid, but in that specific places I do not see how they can be avoided. Feel free to propose a PR if you see a way forward!

**Reply by cymqqqq** [reply]
what do you think about this solution: I changed some input parameters to the reference, and for the insert method(such that btreemap and so on, keep it unchanged), like the code as follows:
For methods such as add_input or something defined in rgb-std, I removed the clone method instead of using the reference.
For the low-level such as the btreemap insert method, I kept the input parameters unchanged. (Because it needs to implement lifetime parameters if we want to use the copy trait here, otherwise we must use clone method)
origin => main_builder = main_builder.add_input(opout, state.clone())?;  // in rgb-std/src/persistent/inventory.rs. line 784
now => main_builder = main_builder.add_input(opout, &state)?;

for the add_input method, the changed code:
pub fn add_input(mut self, opout: Opout, state: &TypedState) -> Result<Self, BuilderError> {
        self.inputs.insert(Input::with(opout), state.clone())?;
        Ok(self)
    }

**Reply by dr-orlovsky** [reply]
The question is will it work with a rust borrow checker :) Let's try
